pkg/function/restore_csi_snapshot.go

```diff
@@ -0,0 +1,161 @@
+	RestoreCSISnapshotRestoreSizeArg = "restoreSize"
+	// RestoreCSISnapshotAccessModesArg lists down the accessmodes for the underlying PV
+	RestoreCSISnapshotAccessModesArg = "accessModes"
+	// RestoreCSISnapshotLabelsArg has labels that will be added to the restored PVC
+	RestoreCSISnapshotLabelsArg = "labels"
+	// RestoreCSISnapshotVolumeModeArg defines mode of volume
+	RestoreCSISnapshotVolumeModeArg = "volumeMode"
+)
+
+type restoreCSISnapshotFunc struct{}
+
+type restoreCSISnapshotArgs struct {
+	Name         string
+	PVC          string
+	Namespace    string
+	StorageClass string
+	RestoreSize  *resource.Quantity
+	AccessModes  []v1.PersistentVolumeAccessMode
+	Labels       map[string]string
+	VolumeMode   v1.PersistentVolumeMode
+}
+
+func (*restoreCSISnapshotFunc) Name() string {
+	return RestoreCSISnapshotFuncName
+}
+
+func (*restoreCSISnapshotFunc) Exec(ctx context.Context, tp param.TemplateParams, args map[string]interface{}) (map[string]interface{}, error) {
+	var restoreSize string
+	var restoreArgs restoreCSISnapshotArgs
+	if err := Arg(args, RestoreCSISnapshotNameArg, &restoreArgs.Name); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, RestoreCSISnapshotPVCNameArg, &restoreArgs.PVC); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, RestoreCSISnapshotNamespaceArg, &restoreArgs.Namespace); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, RestoreCSISnapshotStorageClassArg, &restoreArgs.StorageClass); err != nil {
+		return nil, err
+	}
+	if err := Arg(args, RestoreCSISnapshotRestoreSizeArg, &restoreSize); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, RestoreCSISnapshotAccessModesArg, &restoreArgs.AccessModes, []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce}); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, RestoreCSISnapshotVolumeModeArg, &restoreArgs.VolumeMode, v1.PersistentVolumeFilesystem); err != nil {
+		return nil, err
+	}
+	if err := OptArg(args, RestoreCSISnapshotLabelsArg, &restoreArgs.Labels, nil); err != nil {
+		return nil, err
+	}
+	size := resource.MustParse(restoreSize)
+	restoreArgs.RestoreSize = &size
+
+	if err := restoreCSISnapshot(ctx, restoreArgs); err != nil {
+		return nil, err
+	}
+	return nil, nil
+}
+
+func (*restoreCSISnapshotFunc) RequiredArgs() []string {
+	return []string{
+		RestoreCSISnapshotNameArg,
+		RestoreCSISnapshotPVCNameArg,
+		RestoreCSISnapshotNamespaceArg,
+		RestoreCSISnapshotStorageClassArg,
+		RestoreCSISnapshotRestoreSizeArg,
+	}
+}
+
+func restoreCSISnapshot(ctx context.Context, args restoreCSISnapshotArgs) error {
+	kubeCli, err := kube.NewClient()
+	if err != nil {
+		return err
+	}
+	pvc := getPVCManifest(args)
+	if _, err = kubeCli.CoreV1().PersistentVolumeClaims(args.Namespace).Create(ctx, &pvc, metav1.CreateOptions{}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func getPVCManifest(args restoreCSISnapshotArgs) v1.PersistentVolumeClaim {
```

Comment:
```suggestion
func newPVCManifest(args restoreCSISnapshotArgs) *v1.PersistentVolumeClaim {
```
Since we are creating a new manifest and not getting an existing one from args. Probably okay to return a pointer. Change it back if it does not work